rest-api/src/main/java/com/cloudbees/workflow/rest/external/ChunkVisitor.java

```diff
@@ -0,0 +1,130 @@
+package com.cloudbees.workflow.rest.external;
+
+import com.google.common.collect.Iterables;
+import org.jenkinsci.plugins.workflow.actions.ErrorAction;
+import org.jenkinsci.plugins.workflow.actions.NotExecutedNodeAction;
+import org.jenkinsci.plugins.workflow.actions.TimingAction;
+import org.jenkinsci.plugins.workflow.graph.BlockEndNode;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+import org.jenkinsci.plugins.workflow.graphanalysis.ForkScanner;
+import org.jenkinsci.plugins.workflow.graphanalysis.MemoryFlowChunk;
+import org.jenkinsci.plugins.workflow.graphanalysis.StandardChunkVisitor;
+import org.jenkinsci.plugins.workflow.job.WorkflowRun;
+import org.jenkinsci.plugins.workflow.pipelinegraphanalysis.GenericStatus;
+import org.jenkinsci.plugins.workflow.pipelinegraphanalysis.StatusAndTiming;
+import org.jenkinsci.plugins.workflow.pipelinegraphanalysis.TimingInfo;
+import org.jenkinsci.plugins.workflow.support.actions.PauseAction;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collection;
+
+/**
+ * Couples to the new analysis APIs to collect stages for processing.
+ * This is where all the interesting parts happen.
+ * @author Sam Van Oort
+ */
+public class ChunkVisitor extends StandardChunkVisitor {
+    ArrayDeque<StageNodeExt> stages = new ArrayDeque<StageNodeExt>();
+    FlowNode firstExecuted = null;
+    ArrayDeque<AtomFlowNodeExt> stageContents = new ArrayDeque<AtomFlowNodeExt>();
+    WorkflowRun run;
+
+    public ChunkVisitor(@Nonnull WorkflowRun run) {
+        this.run = run;
+    }
+
+    public Collection<StageNodeExt> getStages() {
+        return stages;
+    }
+
+    public static AtomFlowNodeExt makeAtomNode(@Nonnull WorkflowRun run, @CheckForNull FlowNode beforeNode, @Nonnull FlowNode node, @CheckForNull FlowNode next) {
+        long pause = PauseAction.getPauseDuration(node);
+        TimingInfo times = StatusAndTiming.computeChunkTiming(run, pause, node, node, next);
+        ExecDuration dur = (times == null) ? new ExecDuration() : new ExecDuration(times);
+
+        GenericStatus status = StatusAndTiming.computeChunkStatus(run, beforeNode, node, node, next);
+        if (status == null) {
+            status = GenericStatus.NOT_EXECUTED;
+        }
+        ErrorAction err = node.getError();
+        if (status == GenericStatus.FAILURE && err == null) {
+            // Needed for FlowEndNode, since an ErrorAction isn't generated
+            // When build result is set directly
+            err = new ErrorAction(new Throwable("Run failed, assuming last chunk failed too!"));
```

Comment:
:ant: IIUC this would become a user-visible message, so its terminology should be chosen more carefully. What is a “chunk” to a user?

Perhaps

```java
new AbortException("Build marked as a failure.");
```